File: backend/src/utils/db_init.py
import logging
import os
import time
from pathlib import Path
from backend.seeds.populate_user_data import run_seeds

import psycopg2

logger = logging.getLogger(__name__)

BACKEND_DIR = Path(__file__).resolve().parent.parent.parent
MIGRATIONS_DIR = BACKEND_DIR / "migrations"


def _get_connection():
    return psycopg2.connect(
        host=os.getenv("PGHOST", "localhost"),
        port=os.getenv("PGPORT", "5432"),
        dbname=os.getenv("PGDATABASE", "user_data"),
        user=os.getenv("PGUSER", "postgres"),
        password=os.getenv("PGPASSWORD", "postgres"),
    )


def run_migrations() -> bool:
    try:
        conn = _get_connection()
        conn.autocommit = True
        cur = conn.cursor()
        for sql_file in sorted(MIGRATIONS_DIR.glob("*.sql")):
            logger.info("Running migration: %s", sql_file.name)
            cur.execute(sql_file.read_text())
        cur.close()
        conn.close()
        print("Migrations completed.")
        return True
    except psycopg2.OperationalError as e:
        print(f"Migrations skipped (database not available): {e}")
        return False


def init_db():
    max_attempts = 10
    for attempt in range(max_attempts):
        try:
            conn = _get_connection()
            conn.close()
            break
        except psycopg2.OperationalError:
            if attempt < max_attempts - 1:
                logger.info("Waiting for database... (attempt %d/%d)", attempt + 1, max_attempts)
                time.sleep(2)
            else:
                logger.warning("Database not available after %d attempts. Skipping init.", max_attempts)
                return

    if run_migrations():

        run_seeds()

File: backend/src/utils/db_init.js
import { fileURLToPath, pathToFileURL } from "url";
import { dirname, join } from "path";
import fs from "fs";
import { pool } from "../middleware/database.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const BACKEND_DIR = join(__dirname, "..", "..");
const MIGRATIONS_DIR = join(BACKEND_DIR, "migrations");
const SEEDS_DIR = join(BACKEND_DIR, "seeds");

export async function runMigrations() {
  try {
    const sqlFiles = fs.readdirSync(MIGRATIONS_DIR).filter((f) => f.endsWith(".sql")).sort();
    const client = await pool.connect();
    try {
      await client.query(`
        CREATE TABLE IF NOT EXISTS migrations_run (
          filename VARCHAR(255) PRIMARY KEY
        );
      `);

      for (const sqlFile of sqlFiles) {
        const existsResult = await client.query(
          "SELECT 1 FROM migrations_run WHERE filename = $1",
          [sqlFile]
        );
        if (existsResult.rows.length > 0) {
          console.log(`Skipping migration (already run): ${sqlFile}`);
          continue;
        }

        const sqlPath = join(MIGRATIONS_DIR, sqlFile);
        const sql = fs.readFileSync(sqlPath, "utf-8");
        await client.query(sql);
        await client.query("INSERT INTO migrations_run (filename) VALUES ($1)", [sqlFile]);
        console.log(`Running migration: ${sqlFile}`);
      }
      console.log("Migrations completed.");
      return true;
    } finally {
      client.release();
    }
  } catch (e) {
    console.log(`Migrations skipped (database not available): ${e.message}`);
    return false;
  }
}

export async function runSeeds() {
  try {
    const seedFiles = fs.readdirSync(SEEDS_DIR).filter((f) => f.endsWith(".js")).sort();
    const client = await pool.connect();
    try {
      for (const seedFile of seedFiles) {
        const existsResult = await client.query(
          "SELECT 1 FROM seeds_run WHERE script_name = $1",
          [seedFile]
        );
        if (existsResult.rows.length > 0) {
          console.log(`Seeding skipped (already run): ${seedFile}`);
          continue;
        }

        const seedPath = join(SEEDS_DIR, seedFile);
        const mod = await import(pathToFileURL(seedPath).href);
        const run = mod.default ?? mod.run;
        if (typeof run !== "function") {
          console.log(`Seeding skipped (no default/run export): ${seedFile}`);
          continue;
        }

        const success = await run();
        if (success) {
          await client.query("INSERT INTO seeds_run (script_name) VALUES ($1)", [seedFile]);
          console.log(`Seeding completed: ${seedFile}`);
        }
      }
    } finally {
      client.release();
    }
  } catch (e) {
    console.log(`Seeding error: ${e.message}`);
  }
}

export async function initDb() {
  // When Docker Postgres starts slower than Node, avoid crashing: retry connect, then migrate/seed.
  const maxAttempts = 10;
  for (let attempt = 0; attempt < maxAttempts; attempt++) {
    try {
      const client = await pool.connect();
      client.release();
      break;
    } catch {
      if (attempt < maxAttempts - 1) {
        console.log(`Waiting for database... (attempt ${attempt + 1}/${maxAttempts})`);
        await new Promise((r) => setTimeout(r, 2000));
      } else {
        console.log(`Database not available after ${maxAttempts} attempts. Skipping init.`);
        return;
      }
    }
  }

  await runMigrations();
  await runSeeds();
}
